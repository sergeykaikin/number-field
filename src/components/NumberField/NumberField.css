.number-field {
  border: solid 1px #ccc;
  background-color: #ddd;
  width: 200px;
}

.number-field .number-field__label {
  display: block;
  height: 25px;
  line-height: 25px;
  text-align: center;
}

.number-field .number-field__input {
  display: inline-block;
  vertical-align: bottom;
  border: none;
  border-radius: 0px;
  padding: 0px;
  width: calc(200px - 40px);
  height: 20px;
  text-align: center;
}

.number-field .number-field__button {
  user-select: none;
  cursor: pointer;
  display: inline-block;
  vertical-align: bottom;
  border: 0;
  background-color: #eee;
  height: 20px;
  width: 20px;
  padding: 0;
  line-height: 20px;
  text-align: center;
  font-weight: bold;
}
.number-field .number-field__button.number-field__button:disabled {
  background: transparent;
  color: gray;
  cursor: not-allowed;
}
