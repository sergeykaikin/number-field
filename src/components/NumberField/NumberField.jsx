import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import { getUniqueId } from '../../utils';

import './NumberField.css';

const NumberField = ({ label, value, onChange }) => {
  const idSuffix = useRef(getUniqueId());
  const [number, setNumber] = useState(value);
  const updateValue = value => {
    setNumber(value);
    onChange(value);
  };
  const handleChange = event => void updateValue(Number(event.target.value));
  const handleDecrease = () => void updateValue(number - 1);
  const handleIncrease = () => void updateValue(number + 1);
  const shouldPreventInput = newValue => {
    const newNumberValue = Number(newValue);

    if (isNaN(newNumberValue) || String(newValue).length !== newNumberValue.toFixed(0).length || newNumberValue < 0) {
      return true;
    }

    return false;
  }
  const handleKeyPress = (event) => {
    const { value, selectionStart, selectionEnd } = event.target;
    const { key: inputChar } = event;

    if (shouldPreventInput(`${value.slice(0, selectionStart)}${inputChar}${value.slice(selectionEnd, value.length)}`)) {
      event.preventDefault();
    }
  }
  const handlePaste = event => {
    if (shouldPreventInput(event.clipboardData.getData('Text'))) {
      event.preventDefault();
    }
  }

  return (
    <div className="number-field">
      <label className="number-field__label" htmlFor={`number-field-${idSuffix.current}`}>{label}</label>
      <button
        className="number-field__button number-field__button--minus"
        disabled={number === 0}
        type="button"
        onClick={number === 0 ? undefined : handleDecrease}
      >
        -
      </button>
      <input
        className="number-field__input"
        id={`number-field-${idSuffix.current}`}
        type="text"
        min="0"
        inputMode="numeric"
        pattern="[0-9]*"
        value={number}
        onKeyPress={handleKeyPress}
        onPaste={handlePaste}
        onChange={handleChange}
      />
      <button
        className="number-field__button number-field__button--plus"
        type="button"
        onClick={handleIncrease}
      >
        +
      </button>
    </div>
  )
};

NumberField.defaultProps = {
  value: 0,
  onChange: undefined
};

NumberField.propTypes = {
  label: PropTypes.string.isRequired,
  value: PropTypes.number,
  onChange: PropTypes.func
};

export default React.memo(NumberField);
