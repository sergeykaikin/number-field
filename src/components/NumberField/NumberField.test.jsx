import React from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import NumberField from '.';

describe('NumberField', () => {
  it('"minus" button disabled', () => {
    const tree = renderer
      .create(<NumberField label="test" value={0} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('"minus" and "plus" buttons enabled', () => {
    const tree = renderer
      .create(<NumberField label="test" value={101} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should trigger onChange callback on "plus" and "minus" button clicks', () => {
    const onChange = jest.fn();
    let container = document.createElement('div');
    document.body.appendChild(container);
    act(() => {
      ReactDOM.render(<NumberField label="test" value={101} onChange={onChange} />, container);
    });

    act(() => {
      container
        .querySelector('.number-field__button--minus')
        .dispatchEvent(new MouseEvent('click', {bubbles: true}));
    });
    act(() => {
      container
        .querySelector('.number-field__button--plus')
        .dispatchEvent(new MouseEvent('click', {bubbles: true}));
    });

    expect(onChange.mock.calls.length).toBe(2);
    expect(onChange.mock.calls[0][0]).toBe(100);
    expect(onChange.mock.calls[1][0]).toBe(101);

    document.body.removeChild(container);
    container = null;
  });
});
