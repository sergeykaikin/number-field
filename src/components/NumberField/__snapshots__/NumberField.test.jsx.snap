// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberField "minus" and "plus" buttons enabled 1`] = `
<div
  className="number-field"
>
  <label
    className="number-field__label"
    htmlFor="number-field-2"
  >
    test
  </label>
  <button
    className="number-field__button number-field__button--minus"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    -
  </button>
  <input
    className="number-field__input"
    id="number-field-2"
    inputMode="numeric"
    min="0"
    onChange={[Function]}
    onKeyPress={[Function]}
    onPaste={[Function]}
    pattern="[0-9]*"
    type="text"
    value={101}
  />
  <button
    className="number-field__button number-field__button--plus"
    onClick={[Function]}
    type="button"
  >
    +
  </button>
</div>
`;

exports[`NumberField "minus" button disabled 1`] = `
<div
  className="number-field"
>
  <label
    className="number-field__label"
    htmlFor="number-field-1"
  >
    test
  </label>
  <button
    className="number-field__button number-field__button--minus"
    disabled={true}
    type="button"
  >
    -
  </button>
  <input
    className="number-field__input"
    id="number-field-1"
    inputMode="numeric"
    min="0"
    onChange={[Function]}
    onKeyPress={[Function]}
    onPaste={[Function]}
    pattern="[0-9]*"
    type="text"
    value={0}
  />
  <button
    className="number-field__button number-field__button--plus"
    onClick={[Function]}
    type="button"
  >
    +
  </button>
</div>
`;
